Sources/NIOHTTP2/StreamMap.swift

```diff
@@ -220,6 +243,51 @@ extension CircularBuffer where Element: PerStreamData {
 
         return nil
     }
+
+    fileprivate func findIndexForFirstStreamIDLargerThan(_ streamID: HTTP2StreamID) -> Index {
+        if self.count < binarySearchThreshold {
+            return self.linearScanForFirstStreamIDLargerThan(streamID)
+        } else {
+            return self.binarySearchForFirstStreamIDLargerThan(streamID)
+        }
+    }
+
+    private func linearScanForFirstStreamIDLargerThan(_ streamID: HTTP2StreamID) -> Index {
+        var index = self.startIndex
+
+        while index < self.endIndex {
+            let currentStreamID = self[index].streamID
+            if currentStreamID > streamID {
+                return index
+            }
+            self.formIndex(after: &index)
+        }
+
+        return self.endIndex
+    }
+
+    // Binary search is somewhat complex code compared to a linear scan, so we don't want to inline this code if we can avoid it.
+    @inline(never)
+    private func binarySearchForFirstStreamIDLargerThan(_ streamID: HTTP2StreamID) -> Index {
+        var left = self.startIndex
+        var right = self.endIndex
+
+        while left != right {
+            let pivot = self.index(left, offsetBy: self.distance(from: left, to: right) / 2)
+            let currentStreamID = self[pivot].streamID
+            if currentStreamID > streamID {
+                right = pivot
+            } else if currentStreamID == streamID {
+                // The stream ID is the one after.
+                return self.index(after: pivot)
+            } else {
+                left = self.index(after: pivot)
+            }
+        }
+
+        // ?
```

Comment:
Not a confidence inspiring comment!

Reply:
Hah! I left that for myself when I was unsure if the logic was quite right. I convinced myself it was.